include/tvm/runtime/container.h

```diff
@@ -188,6 +188,709 @@ class InplaceArrayBase {
+    ObjectRef* iter = p->MutableBegin();
+    if (data_.unique()) {
+      for (int64_t i = 0; i < p->size_; ++i) {
+        T old_elem = DowncastNoCheck<T>(std::move(*iter));
+        T new_elem = fmutate(std::move(old_elem));
+        *iter++ = std::move(new_elem);
+      }
+    } else {
+      for (int64_t i = 0; i < p->size_; ++i) {
+        T old_elem = DowncastNoCheck<T>(*iter);
+        T new_elem = fmutate(old_elem);
+        // do nothing until the first real mutation begins.
+        if (new_elem.same_as(*iter++)) {
+          continue;
+        }
+        ObjectPtr<ArrayNode> copy = ArrayNode::make(p->capacity_);
+        // copy [0, i)
+        int64_t& size = copy->size_ = 0;
+        for (ObjectRef* old_iter = p->MutableBegin(); size < i;) {
+          copy->EmplaceInit(size++, *old_iter++);
+        }
+        // copy [i]
+        copy->EmplaceInit(size++, std::move(new_elem));
+        // complete transformation in (i, size)
+        while (size < p->size_) {
+          T old_elem = DowncastNoCheck<T>(*iter++);
+          T new_elem = fmutate(old_elem);
+          copy->EmplaceInit(size++, std::move(new_elem));
+        }
+        data_ = std::move(copy);
+        break;
+      }
+    }
+  }
+
+  /*!
+   * \brief reset the array to content from iterator.
+   * \param begin begin of iterator
+   * \param end end of iterator
+   * \tparam IterType The type of iterator
+   */
+  template <typename IterType>
+  inline void Assign(IterType begin, IterType end) {
+    const int64_t cap = std::distance(begin, end);
+    CHECK_GE(cap, 0) << "ValueError: cannot construct an Array of negative size";
+    ObjectPtr<ArrayNode> p = ArrayNode::make(cap);
+    for (int64_t& i = p->size_ = 0; begin != end; ++i, ++begin) {
+      p->EmplaceInit(i, *begin);
+    }
+    data_ = std::move(p);
+  }
+
+  /*! \brief specify container node */
+  using ContainerType = ArrayNode;
+
+ private:
+  /*!
+   * \brief Move or copy the ArrayNode to new address with the given capacity
+   * \param capacity The capacity requirement of the new address
+   */
+  inline void SwitchContainer(const int64_t capacity) {
+    CHECK_GE(capacity, 0);
+    ObjectPtr<ArrayNode> p = ArrayNode::make(capacity);
+    if (data_ != nullptr) {
+      ArrayNode* data = GetArrayNode();
+      int64_t size = data->size_;
+      ObjectRef* from = data->MutableBegin();
+      ObjectRef* to = p->MutableBegin();
+      if (data_.unique()) {
+        for (int64_t& i = p->size_ = 0; i < size; ++i) {
+          new (to++) ObjectRef(std::move(*from++));
+        }
+      } else {
+        for (int64_t& i = p->size_ = 0; i < size; ++i) {
+          new (to++) ObjectRef(*from++);
+        }
+      }
+    }
+    data_ = std::move(p);
+  }
+
+  /*!
+   * \brief Move elements in [idx, size) backward by `numel`
+   * \param idx The position of the first element to be moved
+   * \param numel The distance for the moving
+   */
+  inline void MoveBackwards(const int64_t idx, const int64_t numel) {
```

Comment:
Backwards is a bit confusion. How about break it into Enlarge, and then MoveElements(like memmove)?

Reply:
The extra cost here is that enlarge initialized the new memory with nulls, but it shouldn't be too much overhead imo

Reply:
The problem of memmove-like method is that it still has to take care of source-destination overlaps. If an extra buffer is not introduced, it still requires us to discuss the two cases (move forward and move backward)

Reply:
i agree we could have two memmove like method in the ArrayNode. I just think term "foward/backward" is a bit confusing . if it is not bottleneck, perhaps we could just use if to switch between the two cases and use a single function name. Or come up with a good name for the two function